rules/firestore.json

```diff
@@ -35,50 +37,204 @@ service cloud.firestore {
+            allow update: if validJoinerUpdate() &&
+                             preservesField('chosenOpponent') &&
+                             acceptsConfig()
+            //   - the chosen opponent can remove itself from the corresponding field
+            allow update: if validJoinerUpdate() &&
+                             preservesField('partStatus') &&
+                             removesOneSelfFromChosenOpponent();
+            // Any verified user can read a joiner
+            allow read: if userIsVerified();
+        }
+        function validJoinerUpdate() {
+            return userIsVerified() &&
+                userIsCreator(resource.data) == false &&
+                userIsChosenOpponent() &&
+                preservesField('creator') &&
+                preservesField('firstPlayer') &&
+                preservesField('partType') &&
+                preservesField('maximalMoveDuration') &&
+                preservesField('totalPartDuration');
+        }
+        function joinerHasNotStarted(joiner) {
+            return joiner.partStatus < 3;
+        }
+        match /joiners/{partId}/candidates/{userId} {
+            // Users can only add themselves to the candidate list if they are not the creator and if they use the right username
+            // We also allow updates in this way, in case someone tries to join the part twice
+            allow create, update: if userIsVerified() && userId == request.auth.uid &&
+                             request.auth.uid != creatorOf(partId).id &&
+                             request.resource.data.name == usernameOf(request.auth.uid) &&
+                             request.resource.data.id == request.auth.uid;
+            // Users can only delete themselves
+            allow delete: if userIsVerified() && userId == request.auth.uid;
+            // Any verified user can read the candidate list
+            allow read: if userIsVerified();
+        }
+        function creatorOf(partId) {
+            return get(/databases/$(database)/documents/joiners/$(partId)).data.creator;
+        }
+        function userIsVerified() {
+            return request != null &&
+                   request.auth != null &&
                    request.auth.uid != null &&
                    request.auth.token.email_verified == true;
         }
-        function isAuthentified() {
-            return request != null &&
-                   request.auth != null;
+        function userIsCreator(joiner) {
+            return joiner.creator != null &&
+                   joiner.creator.id == request.auth.uid;
         }
-        function isOwner(userId) {
-            return request != null &&
-                request.auth != null &&
-                request.auth.uid == userId;
+        function userIsCreatorWithValidName(joiner) {
+            return userIsCreator(joiner) &&
+                   joiner.creator.name == usernameOf(request.auth.uid);
+        }
+        function preservesField(field) {
```

Comment:
fieldName :p ? (autism is strong with this one)

Reply:
changé ! :d